arex-schedule-web-api/src/main/java/com/arextest/schedule/service/ReplayCaseTransmitServiceImpl.java

```diff
@@ -190,6 +192,100 @@ private void doSendValuesToRemoteHost(List<ReplayActionCaseItem> values,
+            ServiceUrlUtils.getHost(targetServiceInstance.getUrl()) : targetServiceInstance.getIp();
+    if (StringUtils.isBlank(host)) {
+      LOGGER.warn("The distribution task failed to start due to an empty ip being provided.");
+      return;
+    }
+    LOGGER.info("start doDistribute to remote host:{}", host);
+    SendSemaphoreLimiter currentLimiter = planExecutionContext.getPlan().getRateLimiterFactory().get(host);
+    if (currentLimiter == null) {
+      LOGGER.warn("The current service instance - [{}],has no rate limiter,skip send.", host);
+      return;
+    }
+
+    List<ReplayActionCaseItem> caseItems = new ArrayList<>(currentLimiter.getPermits());
+
+    while (!caseItemArrayBlockingQueue.isEmpty()) {
+      if (currentLimiter.failBreak()) {
+        // remove interrupted target instance
+        planExecutionContext.removeTargetInstance(targetServiceInstance);
+        LOGGER.warn("The current service instance - [{}],has reached the interruption threshold and the task has been stoped.",
+                host);
+        break;
+      }
+      caseItemArrayBlockingQueue.drainTo(caseItems, currentLimiter.getPermits());
+
+      caseItems.forEach(caseItem->
+              this.doExecute(caseItem, targetServiceInstance, groupSentLatch, planExecutionContext, currentLimiter));
+      caseItems.clear();
+    }
+  }
+
+  private void setServiceInstance(ReplayActionCaseItem replayActionCaseItem,
+                                  ServiceInstance targetServiceInstance,
+                                  PlanExecutionContext<?> planExecutionContext) {
+    replayActionCaseItem.setTargetInstance(targetServiceInstance);
+
+    Map<ServiceInstance, List<ServiceInstance>> bindMap = planExecutionContext.getBindInstanceMap();
+    if (bindMap == null || bindMap.isEmpty()) {
+      return;
+    }
+    List<ServiceInstance> sourceInstances = bindMap.get(targetServiceInstance);
+    if (CollectionUtils.isNotEmpty(sourceInstances)) {
+      int index = Math.abs(replayActionCaseItem.getId().hashCode() % sourceInstances.size());
+      replayActionCaseItem.setSourceInstance(sourceInstances.get(index));
+      LOGGER.info("The source instance is set to [{}], the target instance is set to [{}].",
+              replayActionCaseItem.getSourceInstance().getIp(), replayActionCaseItem.getTargetInstance().getIp());
+    }
+  }
+
+  private void doExecute(ReplayActionCaseItem replayActionCaseItem,
+                         ServiceInstance targetServiceInstance,
+                         CountDownLatch groupSentLatch,
+                         PlanExecutionContext<?> executionContext,
+                         SendSemaphoreLimiter sendSemaphoreLimiter) {
+
+    ReplayActionItem actionItem = replayActionCaseItem.getParent();
+    MDCTracer.addDetailId(replayActionCaseItem.getId());
+
+    try {
+      ReplaySender replaySender = findReplaySender(replayActionCaseItem);
+      if (replaySender == null) {
+        groupSentLatch.countDown();
```

Comment:
if following code throw exception, `groupSentLatch.countDown()` may execute twice?

Reply:
The following method 'doSendFailedAsFinish' handles exceptions without throwing any exceptions